src/platforms/common/stm32/traceswoasync.c

```diff
@@ -106,35 +136,90 @@ void traceswo_setspeed(uint32_t baudrate)
+	if (dma_get_interrupt_flag(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_HTIF)) {
+		dma_clear_interrupt_flags(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_HTIF);
 		memcpy(&trace_rx_buf[write_index * TRACE_ENDPOINT_SIZE], pingpong_buf, TRACE_ENDPOINT_SIZE);
 	}
-	if (DMA_ISR(SWO_DMA_BUS) & DMA_ISR_TCIF(SWO_DMA_CHAN)) {
-		DMA_IFCR(SWO_DMA_BUS) |= DMA_ISR_TCIF(SWO_DMA_CHAN);
+	if (dma_get_interrupt_flag(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_TCIF)) {
+		dma_clear_interrupt_flags(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_TCIF);
 		memcpy(
 			&trace_rx_buf[write_index * TRACE_ENDPOINT_SIZE], &pingpong_buf[TRACE_ENDPOINT_SIZE], TRACE_ENDPOINT_SIZE);
 	}
 	write_index = (write_index + 1U) % NUM_TRACE_PACKETS;
 	trace_buf_drain(usbdev, TRACE_ENDPOINT | USB_REQ_TYPE_IN);
 }
+#else
+void SWO_DMA_ISR(void)
+{
+	if (dma_get_interrupt_flag(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_HTIF)) {
+		dma_clear_interrupt_flags(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_HTIF);
+		memcpy(&trace_rx_buf[write_index * TRACE_ENDPOINT_SIZE], pingpong_buf, TRACE_ENDPOINT_SIZE);
+	}
+	if (dma_get_interrupt_flag(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_TCIF)) {
+		dma_clear_interrupt_flags(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_TCIF);
+		memcpy(
+			&trace_rx_buf[write_index * TRACE_ENDPOINT_SIZE], &pingpong_buf[TRACE_ENDPOINT_SIZE], TRACE_ENDPOINT_SIZE);
+	}
+	write_index = (write_index + 1U) % NUM_TRACE_PACKETS;
+	trace_buf_drain(usbdev, TRACE_ENDPOINT | USB_REQ_TYPE_IN);
+}
+#endif
 
 void traceswo_init(uint32_t baudrate, uint32_t swo_chan_bitmask)
 {
+#ifdef TRACESWOASYNC_ALLOCS
+	if (trace_rx_buf == NULL) {
+		uint8_t *newbuf = memalign(TRACE_ENDPOINT_SIZE, NUM_TRACE_PACKETS * TRACE_ENDPOINT_SIZE);
+		if (newbuf == NULL) {
+			DEBUG_ERROR("%s: memalign failed, errno=%d\n", __func__, errno);
+			return;
+		}
+		trace_rx_buf = newbuf;
+	}
+#endif
+
 	if (!baudrate)
 		baudrate = SWO_DEFAULT_BAUD;
 
 	rcc_periph_clock_enable(SWO_UART_CLK);
 	rcc_periph_clock_enable(SWO_DMA_CLK);
 
+#if defined(STM32F1)
 	gpio_set_mode(SWO_UART_PORT, GPIO_MODE_INPUT, GPIO_CNF_INPUT_PULL_UPDOWN, SWO_UART_RX_PIN);
 	/* Pull SWO pin high to keep open SWO line ind uart idle state! */
 	gpio_set(SWO_UART_PORT, SWO_UART_RX_PIN);
+#else
+	gpio_mode_setup(SWO_UART_PORT, GPIO_MODE_AF, GPIO_PUPD_PULLUP, SWO_UART_RX_PIN);
+	gpio_set_output_options(SWO_UART_PORT, GPIO_OTYPE_OD, GPIO_OSPEED_100MHZ, SWO_UART_RX_PIN);
+	gpio_set_af(SWO_UART_PORT, GPIO_AF7, SWO_UART_RX_PIN);
+#endif
+
 	nvic_set_priority(SWO_DMA_IRQ, IRQ_PRI_SWO_DMA);
 	nvic_enable_irq(SWO_DMA_IRQ);
 	traceswo_setspeed(baudrate);
 	traceswo_setmask(swo_chan_bitmask);
 	decoding = (swo_chan_bitmask != 0);
 }
+
+void traceswo_deinit(void)
+{
+	/* Stop peripherals servicing */
+	nvic_disable_irq(SWO_DMA_IRQ);
+	dma_disable_channel(SWO_DMA_BUS, SWO_DMA_CHAN);
+	usart_disable(SWO_UART);
+	/* Dump the buffered remains */
+	trace_buf_drain(usbdev, TRACE_ENDPOINT | USB_REQ_TYPE_IN);
+#ifdef TRACESWOASYNC_DEALLOCS
+	/* Release this chunk of precious SRAM. FIXME: does heap shrink after free/malloc_trim? */
```

Comment:
Heap never shrinks under newlib-nano, but it's also not helpful or useful to either worry about or force - when the allocation is `free()`'d back to the allocation pool, it will automatically be re-used by future allocations, either whole or in parts. -nano's allocator cannot un-`sbrk()` memory and will aggressively re-use chunks in the free list in preference for `sbrk()`'ing new memory, as long as it can find enough contiguous spare in the list.

Reply:
Yes, by now I know newlib-nano `free` does no trimming and there's no `malloc_trim` on everything F1 in-tree. I wanted to make space for possible (descending) stack growth, as some Big things in BMF are allocated on stack. This is less relevant in light of PR1830. I see how sbrk is forward-only, and it will eventually reach end-of-SRAM (not MSP stack), and after a traceswo_deinit, any subsequent `malloc` may pick this space.
The code might be useful where newlib is available (F4), and in out-of-tree projects (with RTOS and complex stuff). If this were NuttX, I might've asked to allocate in DMA-able memory (e.g. not DTCM).
I can remove the comments, but I don't think the code itself needs to go. Are there nano-specific feature macros I could test this against? `mallopt` has to go before this `free`, might go in initialisation fwiw. `malloc_trim` has to be hidden.

Reply:
Dropped code incompatible with nano, reworded comment.